Tokenization tutorial (#5257)
huggingface/transformers: docs/source/preprocessing.rst

Preprocessing data
==================

> In this tutorial, we'll explore how to preocess your data using 🤗 Transformers. The main tool for this is what we

Reviewer: ```suggestion
In this tutorial, we'll explore how to preprocess your data using 🤗 Transformers. The main tool for this is what we

```

is its ``__call__``: you just need to feed your sentence to your tokenizer object.

::

    encoded_input = tokenizer("Hello, I'm a single sentence!")
    print(encoded_input)

This will return a dictionary string to list of ints like this one:

::

    {'input_ids': [101, 138, 18696, 155, 1942, 3190, 1144, 1572, 13745, 1104, 159, 9664, 2107, 102], 
     'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 
     'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

The `input_ids <glossary.html#input-ids>`__ are the indices corresponding to each token in our sentence. We will see
below what the `attention_mask <glossary.html#attention-mask>`__ is used for and in
:ref:`the next section <sentence-pairs>` the goal of `token_type_ids <glossary.html#token-type-ids>`__.

The tokenizer can decode a list of token ids in a proper sentence:

::

    tokenizer.decode(encoded_input["input_ids"])

which should return

::

    "[CLS] Hello, I'm a single sentence! [SEP]"

As you can see, the tokenizer automatically added some special tokens that the model expect. Not all model need special

> tokens; for instance, if we had use `gtp2-medium` instead of `bert-base-cased` to create our tokenizer, we would have

Reviewer: ```suggestion
tokens; for instance, if we had used` gtp2-medium` instead of `bert-base-cased` to create our tokenizer, we would have

```

- To return tensors.

You can do all of this by using the following options when feeding your list of sentences to the tokenizer:

::

    ## PYTORCH CODE
    batch = tokenizer(batch_sentences, padding=True, truncation=True, return_tensors="pt")
    print(batch)
    ## TENSORFLOW CODE
    batch = tokenizer(batch_sentences, padding=True, truncation=True, return_tensors="tf")
    print(batch)

which should now return a dictionary string to tensor like this:

::

    {'input_ids': tensor([[ 101, 8667,  146,  112,  182,  170, 1423, 5650,  102],
                          [ 101, 1262, 1330, 5650,  102,    0,    0,    0,    0],
                          [ 101, 1262, 1103, 1304, 1304, 1314, 1141,  102,    0]]),
     'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0],
                               [0, 0, 0, 0, 0, 0, 0, 0, 0],
                               [0, 0, 0, 0, 0, 0, 0, 0, 0]]), 
     'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1],
                               [1, 1, 1, 1, 1, 0, 0, 0, 0],
                               [1, 1, 1, 1, 1, 1, 1, 1, 0]])}

We can now see what the `attention_mask <glossary.html#attention-mask>`__ is all about: it points out which tokens the

> model should pay attention to and which ones it shoul not (because they represent padding in this case).

Reviewer: ```suggestion
model should pay attention to and which ones it should not (because they represent padding in this case).

```

Note that if your model does not have a maximum length associated to it, the command above will throw a warning. You
can safely ignore it. You can also pass ``verbose=False`` to stop the tokenizer to throw those kinds of warnings.

.. _sentence-pairs:

Preprocessing pairs of sentences
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Sometimes you need to feed pair of sentences to your model. For instance, if you want to classify if two sentences in a
pair are similar, or for question-answering models, which take a context and a question. For BERT models, the input is
then represented like this:

::

    [CLS] Sequence A [SEP] Sequence B [SEP]

> You can encode a pair of sentences in the format your model expects it by passing the two sentences as two arguments

Reviewer: ```suggestion
You can encode a pair of sentences in the format expected by your model by supplying the two sentences as two arguments

```

> (not a list since a list of two sentences will be interpreted as a batch of teo single sentences, as we saw before).

Reviewer: ```suggestion
(not a list since a list of two sentences will be interpreted as a batch of two single sentences, as we saw before).

```

::

    encoded_input = tokenizer("How old are you?", "I'm 6 years old")
    print(encoded_input)

This will once again return a dict string to list of ints:

::

    {'input_ids': [101, 1731, 1385, 1132, 1128, 136, 102, 146, 112, 182, 127, 1201, 1385, 102], 
     'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1], 
     'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

This shows us what the `token_type_ids <glossary.html#token-type-ids>`__ are for: they indicate to the model which part

> of the inputs correspond to the first sentence and which part corresponds to the second sentence.

Reviewer: Should we explain this as well?

The `token_type_ids` are not handled or required by all the models. By default, the tokenizer will only produce inputs that can be handled by the associated model. You can force or prevent the production of various type of inputs with the `return_xxx` arguments.

Author: Good idea!

                       [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}

To double-check what is fed to the model, we can decode each list in `input_ids` one by one:

::

    for ids in encoded_inputs["input_ids"]:
        print(tokenizer.decode(ids))

which will return:

::

    [CLS] Hello I'm a single sentence [SEP] I'm a sentence that goes with the first sentence [SEP]
    [CLS] And another sentence [SEP] And I should be encoded with the second sentence [SEP]
    [CLS] And the very very last one [SEP] And I go with the very last one [SEP]

Once again, you can automatically pad your inputs to the maximum sentence length in the batch, truncate to the maximum
length the model can accept and return tensors directly with the following:

::

    ## PYTORCH CODE
    batch = tokenizer(batch_sentences, batch_of_second_sentences, padding=True, truncation=True, return_tensors="pt")
    ## TENSORFLOW CODE
    batch = tokenizer(batch_sentences, batch_of_second_sentences, padding=True, truncation=True, return_tensors="tf")

Everything you always wanted to know about padding and truncation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> We have seen the command that will work for most cases (pad your batch to the length of the maximum sentence and

Reviewer: ```suggestion
We have seen the commands that will work for most cases (pad your batch to the length of the maximum sentence and

```
